lib/nrf9160_timestamp/nrf9160_timestamp.c

```diff
@@ -0,0 +1,270 @@
+{
+	int err;
+	int i = 0;
+
+	while (i < ARRAY_SIZE(servers)) {
+		err =  sntp_request_time(servers[i].server,
+					 K_SECONDS(5), &sntp_time);
+		if (err) {
+			LOG_DBG("Not getting time from NTP server %s, error %d",
+				log_strdup(servers[i].server), err);
+			LOG_DBG("Trying another address...");
+			i++;
+			continue;
+		}
+
+		LOG_DBG("Got time response from NTP server %s",
+			log_strdup(servers[i].server));
+		time_aux.date_time_utc =
+			(s64_t)sntp_time.seconds * 1000 - k_uptime_get();
+		time_aux.last_date_time_update = k_uptime_get();
+		return 0;
+	}
+
+	LOG_ERR("Not getting time from any NTP server");
+	return -ENODATA;
+}
+
+static int check_current_time(void)
+{
+	int age_last_time_update =
+		k_uptime_get() - time_aux.last_date_time_update;
+	int current_update_time_interval_ms =
+		CONFIG_NRF9160_TIMESTAMP_TIME_UPDATE_INTERVAL * 1000;
+
+	if (time_aux.last_date_time_update == 0 ||
+	    time_aux.date_time_utc == 0) {
+		LOG_DBG("Date time never set");
+		return -ENODATA;
+	}
+
+	if (age_last_time_update > current_update_time_interval_ms) {
+		LOG_DBG("Current date time too old");
+		return -ENODATA;
+	}
+
+	return 0;
+}
+
+void update_new_date_time(void)
+{
+	int err;
+
+	while (true) {
+		k_sem_take(&ntp_sem, K_FOREVER);
+
+		LOG_DBG("Updating date time UTC...");
+
+		err = check_current_time();
+		if (err == 0) {
+			LOG_DBG("Time successfully obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Error getting current time");
+		LOG_DBG("Fallback on cellular network time");
+
+		err = get_time_cellular_network();
+		if (err == 0) {
+			LOG_DBG("Time from cellular network obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Error getting cellular network time");
+		LOG_DBG("Fallback on NTP server");
+
+		err = get_time_NTP_server();
+		if (err == 0) {
+			LOG_DBG("Time from NTP server obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Error getting time from NTP server");
+	}
+}
+
+K_THREAD_DEFINE(ntp_thread, CONFIG_NRF9160_TIME_NTP_THREAD_SIZE,
+		update_new_date_time, NULL, NULL, NULL,
+		K_HIGHEST_APPLICATION_THREAD_PRIO, 0, K_NO_WAIT);
+
+static void nrf9160_time_handler(struct k_work *work)
+{
+	k_sem_give(&ntp_sem);
+
+	LOG_DBG("New date time update in: %d seconds",
+		CONFIG_NRF9160_TIMESTAMP_TIME_UPDATE_INTERVAL);
+
+	k_delayed_work_submit(&time_work.work,
+		K_SECONDS(CONFIG_NRF9160_TIMESTAMP_TIME_UPDATE_INTERVAL));
+}
+
+void nrf9160_time_init(void)
+{
+	strcpy(time_work.name, CONFIG_NRF9160_TIMESTAMP_DEV_NAME);
+
+	k_delayed_work_init(&time_work.work, nrf9160_time_handler);
+	k_delayed_work_submit(&time_work.work, K_NO_WAIT);
+}
+
+void date_time_set(struct tm *new_date_time)
+{
+	time_aux.last_date_time_update = k_uptime_get();
+
+	/* Needed changes so mktime gives UNIX time */
+	new_date_time->tm_year    -= 1900;
+	new_date_time->tm_mon     -= 1;
+
+	time_aux.date_time_utc =
+		(s64_t)mktime(new_date_time) * 1000 - k_uptime_get();
```

Comment:
`mktime()` converts times taking into account the time zone; it does not guarantee to produce UTC.  Consider using the Zephyr function `timeutil_timegm64()` from the upstream `<sys/timeutil.h>` header.

Reply:
Agreed. Changed to ``timeutil_timegm64``